### PR TITLE
Core dump on ignition off when registering 20 apps

### LINKED ISSUE #3857: SDL core dump on Ignition OFF during registering Apps
## Bug Report
SDL core dump on Ignition OFF during registering Apps

##### Occurrences:
~10%

##### Reproduction Steps
1. SDL and HMI are started
2. Connect mobile device via USB transport.
3. Perform ignition Off during registering Apps (25 Apps)
On "HMI" go to → Click the "Exit Application" button → Select “IGNITION_OFF” in “Exit Application reason” → Click the `Send state` button

##### Expected Behavior
SDL starts the ignition off sequence
1. SDL sends to HMI OnSDLClose and stops working
SDL->HMI: BasicCommunication.OnSDLClose:{}

##### Observed Behavior
SDL starts the ignition off sequence
1. SDL sends to HMI OnSDLClose and stops working
SDL->HMI: BasicCommunication.OnSDLClose:{}
Core dump is observed
SIGSEGV caught
DEBUG [139969695552064][SDLMain] /home/jenkins/workspace/Develop_SDL_NoUT_P/src/appMain/life_cycle_impl.cc:241 {anonymous}::sig_handler: SIGSEGV signal has been caught

##### OS & Version Information
* OS/Version: Ubuntu 18
* SDL Core Version:  https://github.com/smartdevicelink/sdl_core/tree/develop (7a4a4e0)

##### Logs:
[Logs.zip](https://github.com/smartdevicelink/sdl_core/files/8061975/Logs.zip)

### PR DESCRIPTION
**Pull request by ShobhitAd**
Fixes #3857 

Original PR: https://github.com/smartdevicelink/sdl_core/pull/3928

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
Submitted ATF test for issue

### Summary
Fixes the following problems:

- Аn object of this class AppExtension was not given externally without protection. Therefore, I will change the return type from Extensions () to DataAccessor <std :: list >. Where the desired mutex is captured.
- ApplicationManagerImpl had unprotected deletion of objects used in other threads. Also, I found a duplicate of the is_stopping method. I deleted this one and left the IsStopping method.
- ResumeCtrlImpl captured the mutex queue_lock_, which belongs to it, and the mutex from ApplicationManagerImpl, when calling the application(..) method. This created a deadlock, and not a secure use of the application object. So I added getting all the apps and working with this list.
- It consists of two different crash of the program. The first crash when remove mobile_request_list_. The second is when deleting mobile requests from waiting_for_response_. This was due to the use of smart pointers in different threads. They are copied into threads, so we have to delete them in the thread in which we finish working with it. Since these requests are processed in a separate thread, I delete them there if the core stops working during the processing of these requests. Also added a check to see if the core stopped working while adding such requests to the list.
- This is a drop while processing data in one thread and deleting it in another thread. I created a separate mutex to protect this data. Because using the existing one, we do not invest in the time allotted to the timer to perform certain work.

### CLA
- [x] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

File: src/components/transport_manager/include/transport_manager/transport_adapter/transport_adapter_impl.h [hunk 1]
```diff
@@ -645,6 +645,8 @@ class TransportAdapterImpl : public TransportAdapter,
    **/
   ConnectionMap connections_;
 
+  mutable sync_primitives::Lock connections_duplicate;
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/components/transport_manager/include/transport_manager/transport_adapter/transport_adapter_impl.h]: You're right, it doesnt seem to make a difference. Removed in https://github.com/smartdevicelink/sdl_core/pull/3939/commits/941ea090672d0cfc6ebf1251cc6c214a27ff1d7b

File: src/components/application_manager/src/resumption/resume_ctrl_impl.cc [hunk 2]
```diff
@@ -252,6 +252,7 @@ bool ResumeCtrlImpl::SetupDefaultHMILevel(ApplicationSharedPtr application) {
 
 void ResumeCtrlImpl::ApplicationResumptiOnTimer() {
   SDL_LOG_AUTO_TRACE();
+  DataAccessor<ApplicationSet> apps = application_manager_.applications();
```
[the remark this answers is not in the document]
Comment (ShobhitAd) [reply src/components/application_manager/src/resumption/resume_ctrl_impl.cc]: Good catch. Removed in https://github.com/smartdevicelink/sdl_core/pull/3939/commits/941ea090672d0cfc6ebf1251cc6c214a27ff1d7b

File: src/components/application_manager/test/request_controller/request_controller_test.cc [hunk 3]
```diff
@@ -211,40 +216,18 @@ TEST_F(RequestControllerTestClass,
...
-
-  // Trying to add one more extra request
-  // Expect overflow and TOO_MANY_PENDING_REQUESTS result
-  EXPECT_EQ(RequestController::TResult::TOO_MANY_PENDING_REQUESTS,
-            AddRequest(settings,
-                       GetMockRequest(),
-                       RequestInfo::RequestType::MobileRequest,
-                       mobile_apis::HMILevel::HMI_FULL));
-}
-
```
Comment (jacobkeeler) [inline src/components/application_manager/test/request_controller/request_controller_test.cc:247]: Do we know why this test was removed?

**Review by jacobkeeler — CHANGES_REQUESTED** [review]

File: src/components/application_manager/src/application_impl.cc [hunk 4]
```diff
@@ -178,12 +178,27 @@ ApplicationImpl::~ApplicationImpl() {
...
     set_perform_interaction_mode(-1);
   }
   CleanupFiles();
+  sync_primitives::AutoLock lock(mobile_message_lock_);
+  mobile_message_queue_.clear();
+  {
+    extensions_lock_->Acquire();
+    extensions_.clear();
+    extensions_lock_->Release();
+  }
```
Comment (jacobkeeler) [inline src/components/application_manager/src/application_impl.cc:201]:
```suggestion
  {
    sync_primitives::AutoLock lock(mobile_message_lock_);
    mobile_message_queue_.clear();
  }
  {
    sync_primitives::AutoLock lock(extensions_lock_);
    extensions_.clear();
    extensions_lock_->Release();
  }
```

File: src/components/application_manager/test/request_controller/request_controller_test.cc (hunk 3 above)
Comment (ShobhitAd) [reply src/components/application_manager/test/request_controller/request_controller_test.cc:247]: That test seems to have been removed due to the change in this line https://github.com/smartdevicelink/sdl_core/pull/3939/files#diff-4f3d0bb78ef3a8c3c3a38cc38dbe26e370440ef9c134c2054f7c0b98b1dd1501R208. The test used request_ctrl_->DestroyThreadpool(); and, due to the added TPoolState::STOPPED != pool_state_  check, no pending requests would be added to the mobile_request_list_ so the number of pending requests wouldn't increase at all (TOO_MANY_PENDING_REQUESTS wouldn't be triggered)

**Review by jacobkeeler — APPROVED** [review]

**Review by jacobkeeler — APPROVED** [review]

**Pull request merged**